File: app/src/types/document.ts
export type Alignment = 'left' | 'center' | 'right' | 'justify';

export type LineSpacingRule = 'auto' | 'exact' | 'at_least';

export type TabStopAlignment = 'left' | 'center' | 'right' | 'decimal';

export type VerticalMerge = 'restart' | 'continue';

export type TableLayout = 'fixed' | 'autofit';

export type PageOrientation = 'portrait' | 'landscape';

export interface NumberingRef {
  num_id: string;
  level: number;
}

export interface TabStop {
  position: number;
  alignment: TabStopAlignment;
  leader?: string;
}

export interface Border {
  style: string;
  size: number;
  color: string;
}

export interface ParagraphProperties {
  alignment?: Alignment;
  heading_level?: number;
  numbering?: NumberingRef;
  spacing_before?: number;
  spacing_after?: number;
  line_spacing?: number;
  line_spacing_rule?: LineSpacingRule;
  indent_left?: number;
  indent_right?: number;
  indent_first_line?: number;
  indent_hanging?: number;
  page_break_before?: boolean;
  keep_next?: boolean;
  keep_lines?: boolean;
  widow_control?: boolean;
  style_id?: string;
  border_bottom?: Border;
  border_top?: Border;
  shading?: string;
  tab_stops?: TabStop[];
}

export interface RunProperties {
  bold?: boolean;
  italic?: boolean;
  underline?: boolean;
  strikethrough?: boolean;
  double_strikethrough?: boolean;
  superscript?: boolean;
  subscript?: boolean;
  font_family?: string;
  font_size?: number;
  color?: string;
  highlight?: string;
  background_color?: string;
  inline_image?: string;
  hyperlink_url?: string;
  hyperlink_tooltip?: string;
  line_break?: boolean;
  page_break?: boolean;
  tab?: boolean;
  small_caps?: boolean;
  all_caps?: boolean;
  spacing?: number;
  footnote_ref?: string;
  endnote_ref?: string;
}

export interface Run {
  id: string;
  text: string;
  properties: RunProperties;
}

export interface Bookmark {
  id: string;
  name: string;
}

export interface Paragraph {
  id: string;
  runs: Run[];
  properties: ParagraphProperties;
  bookmarks?: Bookmark[];
}

export interface TableBorders {
  top?: Border;
  bottom?: Border;
  left?: Border;
  right?: Border;
  inside_h?: Border;
  inside_v?: Border;
}

export interface TableRowProperties {
  height?: number;
  height_rule?: string;
  is_header?: boolean;
  cant_split?: boolean;
}

export interface TableProperties {
  width?: number;
  borders?: TableBorders;
  cell_spacing?: number;
  layout?: TableLayout;
  indent?: number;
}

export interface TableCellProperties {
  width?: number;
  vertical_align?: string;
  borders?: TableBorders;
  shading?: string;
  grid_span?: number;
  vertical_merge?: VerticalMerge;
  text_direction?: string;
  no_wrap?: boolean;
  padding_top?: number;
  padding_bottom?: number;
  padding_left?: number;
  padding_right?: number;
}

export interface TableCell {
  id: string;
  content: BlockElement[];
  properties: TableCellProperties;
}

export interface TableRow {
  id: string;
  cells: TableCell[];
  properties?: TableRowProperties;
}

export interface Table {
  id: string;
  rows: TableRow[];
  properties: TableProperties;
}

export type BlockElement =
  | { paragraph: Paragraph }
  | { table: Table };

export interface ImageData {
  id: string;
  data_base64: string;
  content_type: string;
  width?: number;
  height?: number;
  description?: string;
}

export interface ListLevel {
  level: number;
  format: string;
  text: string;
  start: number;
}

export interface ListDef {
  num_id: string;
  levels: ListLevel[];
}

export interface StyleDef {
  id: string;
  name: string;
  style_type: string;
  based_on?: string;
  run_properties?: RunProperties;
  paragraph_properties?: ParagraphProperties;
}

export interface Header {
  id: string;
  content: BlockElement[];
}

export interface Footer {
  id: string;
  content: BlockElement[];
}

export interface Comment {
  id: string;
  author: string;
  date?: string;
  initials?: string;
  content: BlockElement[];
}

export interface Footnote {
  id: string;
  content: BlockElement[];
}

export interface Endnote {
  id: string;
  content: BlockElement[];
}

export interface SectionProperties {
  page_width?: number;
  page_height?: number;
  margin_top?: number;
  margin_right?: number;
  margin_bottom?: number;
  margin_left?: number;
  margin_header?: number;
  margin_footer?: number;
  margin_gutter?: number;
  page_orientation?: PageOrientation;
  columns?: number;
}

export interface DocumentStatistics {
  word_count: number;
  char_count: number;
  paragraph_count: number;
  page_count_estimate: number;
  image_count: number;
  table_count: number;
  header_count: number;
  footer_count: number;
  style_count: number;
  comment_count: number;
  footnote_count: number;
  endnote_count: number;
}

export interface DocxDocument {
  body: BlockElement[];
  styles: StyleDef[];
  numbering: ListDef[];
  headers: Header[];
  footers: Footer[];
  images: ImageData[];
  comments?: Comment[];
  footnotes?: Footnote[];
  endnotes?: Endnote[];
  section_properties?: SectionProperties;
}

File: crates/docx-wasm/src/model/document.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ---------------------------------------------------------------------------
// Top-level document
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Document {
    pub body: Vec<BlockElement>,
    pub styles: Vec<StyleDef>,
    pub numbering: Vec<ListDef>,
    pub headers: Vec<Header>,
    pub footers: Vec<Footer>,
    pub images: Vec<ImageData>,
}

impl Document {
    /// Returns a default empty document containing a single empty paragraph.
    pub fn new() -> Self {
        let empty_paragraph = Paragraph {
            id: Uuid::new_v4().to_string(),
            runs: Vec::new(),
            properties: ParagraphProperties::default(),
        };

        Self {
            body: vec![BlockElement::Paragraph(empty_paragraph)],
            styles: Vec::new(),
            numbering: Vec::new(),
            headers: Vec::new(),
            footers: Vec::new(),
            images: Vec::new(),
        }
    }
}

impl Default for Document {
    fn default() -> Self {
        Self::new()
    }
}

// ---------------------------------------------------------------------------
// Block-level elements
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "snake_case")]
pub enum BlockElement {
    Paragraph(Paragraph),
    Table(Table),
}

// ---------------------------------------------------------------------------
// Paragraph
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Paragraph {
    pub id: String,
    pub runs: Vec<Run>,
    pub properties: ParagraphProperties,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct ParagraphProperties {
    pub alignment: Option<Alignment>,
    pub heading_level: Option<u8>,
    pub numbering: Option<NumberingRef>,
    pub spacing_before: Option<f64>,
    pub spacing_after: Option<f64>,
    pub indent_left: Option<f64>,
    pub indent_right: Option<f64>,
    pub indent_first_line: Option<f64>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct NumberingRef {
    pub num_id: String,
    pub level: u32,
}

// ---------------------------------------------------------------------------
// Alignment
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "snake_case")]
pub enum Alignment {
    Left,
    Center,
    Right,
    Justify,
}

// ---------------------------------------------------------------------------
// Run (inline content)
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Run {
    pub id: String,
    pub text: String,
    pub properties: RunProperties,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct RunProperties {
    pub bold: Option<bool>,
    pub italic: Option<bool>,
    pub underline: Option<bool>,
    pub strikethrough: Option<bool>,
    pub font_family: Option<String>,
    pub font_size: Option<f64>,
    pub color: Option<String>,
    pub highlight: Option<String>,
    /// Reference to an `ImageData::id` for an inline image.
    pub inline_image: Option<String>,
}

// ---------------------------------------------------------------------------
// Table
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Table {
    pub id: String,
    pub rows: Vec<TableRow>,
    pub properties: TableProperties,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct TableRow {
    pub id: String,
    pub cells: Vec<TableCell>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct TableCell {
    pub id: String,
    pub content: Vec<BlockElement>,
    pub properties: TableCellProperties,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct TableProperties {
    pub width: Option<f64>,
    pub borders: Option<TableBorders>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct TableCellProperties {
    pub width: Option<f64>,
    pub vertical_align: Option<String>,
    pub borders: Option<TableBorders>,
    pub shading: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct TableBorders {
    pub top: Option<Border>,
    pub bottom: Option<Border>,
    pub left: Option<Border>,
    pub right: Option<Border>,
    pub inside_h: Option<Border>,
    pub inside_v: Option<Border>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Border {
    pub style: String,
    pub size: f64,
    pub color: String,
}

// ---------------------------------------------------------------------------
// Images
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ImageData {
    pub id: String,
    pub data_base64: String,
    pub content_type: String,
    pub width: Option<f64>,
    pub height: Option<f64>,
}

// ---------------------------------------------------------------------------
// Numbering / Lists
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ListDef {
    pub num_id: String,
    pub levels: Vec<ListLevel>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ListLevel {
    pub level: u32,
    pub format: String,
    pub text: String,
    pub start: u32,
}

// ---------------------------------------------------------------------------
// Styles
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct StyleDef {
    pub id: String,
    pub name: String,
    pub style_type: String,
    pub based_on: Option<String>,
    pub run_properties: Option<RunProperties>,
    pub paragraph_properties: Option<ParagraphProperties>,
}

// ---------------------------------------------------------------------------
// Headers & Footers
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Header {
    pub id: String,
    pub content: Vec<BlockElement>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Footer {
    pub id: String,
    pub content: Vec<BlockElement>,
}
